packages/benchmark/src/benchmarkerator.js

```diff
@@ -425,43 +439,96 @@ const printBenchmarkStats = stats => {
     // prettier-ignore
     `${stats.rounds} rounds in ${stats.elapsedTime}ns (${stats.timePerRound.toFixed(3)}/round})`,
   );
-  log(`${h1} ${h2} ${h3}`);
-  log(`${d1} ${d2} ${d3}`);
 
-  const data = stats.data;
-  for (const [key, entry] of Object.entries(data)) {
-    const col1 = `${key.padEnd(w1)}`;
-    const col2 = `${String(entry.delta).padStart(w2)}`;
-    const col3 = `${pn(entry.deltaPerRound).padStart(w3)}`;
+  const wc1 = 32;
+  const hc1 = `${'Counter'.padEnd(wc1)}`;
+  const dc1 = `${''.padEnd(wc1, '-')}`;
+
+  const wc2 = 6;
+  const hc2 = `${'Delta'.padStart(wc2)}`;
+  const dc2 = ` ${''.padStart(wc2 - 1, '-')}`;
+
+  const wc3 = 10;
+  const hc3 = `${'PerRound'.padStart(wc3)}`;
+  const dc3 = ` ${''.padStart(wc3 - 1, '-')}`;
+
+  log(``);
+  log(`${hc1} ${hc2} ${hc3}`);
+  log(`${dc1} ${dc2} ${dc3}`);
+  for (const [key, entry] of Object.entries(stats.counters)) {
+    const col1 = `${key.padEnd(wc1)}`;
+    const col2 = `${String(entry.delta).padStart(wc2)}`;
+    const col3 = `${pn(entry.deltaPerRound).padStart(wc3)}`;
     log(`${col1} ${col2} ${col3}`);
   }
+
+  const wg1 = 32;
+  const hg1 = `${'Gauge'.padEnd(wg1)}`;
+  const dg1 = `${''.padEnd(wg1, '-')}`;
+
+  const wg2 = 6;
+  const hg2 = `${'Start'.padStart(wg2)}`;
+  const dg2 = ` ${''.padStart(wg2 - 1, '-')}`;
+
+  const wg3 = 6;
+  const hg3 = `${'End'.padStart(wg3)}`;
+  const dg3 = ` ${''.padStart(wg3 - 1, '-')}`;
+
+  const wg4 = 6;
+  const hg4 = `${'Delta'.padStart(wg4)}`;
+  const dg4 = ` ${''.padStart(wg4 - 1, '-')}`;
+
+  log(``);
+  log(`${hg1} ${hg2} ${hg3} ${hg4}`);
+  log(`${dg1} ${dg2} ${dg3} ${dg4}`);
+  for (const [key, entry] of Object.entries(stats.gauges)) {
+    const col1 = `${key.padEnd(wg1)}`;
+    const col2 = `${String(entry.start).padStart(wg2)}`;
+    const col3 = `${String(entry.end).padStart(wg3)}`;
+    const col4 = `${String(entry.end - entry.start).padStart(wg4)}`;
+    log(`${col1} ${col2} ${col3} ${col4}`);
+  }
 };
 
-const organizeRoundsStats = (
-  rawBefore,
-  rawAfter,
-  elapsedTime,
-  cranks,
-  rounds,
-) => {
+const organizeRoundsStats = (rounds, perRoundStats) => {
+  let elapsedTime = 0;
+  let cranks = 0;
+  for (const { start, end } of perRoundStats) {
+    elapsedTime += Number(end.time - start.time);
+    cranks += end.cranks - start.cranks;
+  }
   const stats = {
     elapsedTime,
     cranks,
     rounds,
     timePerCrank: cranks ? elapsedTime / cranks : 0,
     timePerRound: elapsedTime / rounds,
     cranksPerRound: cranks / rounds,
-    data: {},
+    perRoundStats,
+    counters: {},
+    gauges: {},
   };
 
-  // Note: the following assumes rawBefore and rawAfter have the same keys.
-  for (const [key, value] of Object.entries(rawBefore)) {
+  // Note: the following assumes stats records all have the same keys.
```

Comment:
meaning that `last` has the same keys as `first`? That can be documented and checked by type annoations

Reply:
I'm reluctant to use a type annotation here because this code expressly does not want to have care about what the details actually are, merely that the two things have the same schema.  Is there a generic way to express that?

Reply:
yes. though as I look more closely first and last both are already typed as `rawStats` which is whatever `controller.getStats()` returns and that comes from some code in the kernel.

I don't see how those stats records _could_ have different keys.  Though fwiw, the code here doesn't fail if `last` has _additional_ keys that `first` doesn't; it just ignores them. That's not surprising.

Reply:
Given how the data is produced, there's no way the keys could differ.  It's just since the data is being manipulated blindly I wanted to get the constraint on the record, hence the comment.  Arguably if `last` had additional keys missing from `first` the code would run without crashing but it still might be bug in terms of what was being reported.  In any case, I don't think it's a practical concern right now.